org.eclipse.jdt.core/codeassist/org/eclipse/jdt/internal/codeassist/CompletionEngine.java

```diff
@@ -14801,4 +14778,154 @@ public void findPackagesInModule(char[] prefix,  ModuleBinding module, Scope sco
 		}
 	}
 
+	private void completionAtSwitchCase(ASTNode astNode, SwitchStatement switchStatement, Scope scope,
+			boolean insideTypeAnnotation) {
+		final CompletionOnSingleNameReference singleNameReference = (CompletionOnSingleNameReference) astNode;
+		this.completionToken = singleNameReference.token;
+		if (switchStatement.expression.resolvedType != null) {
+			TypeBinding exprType = switchStatement.expression.resolvedType;
+			// handle completions of enum literals
+			if (exprType.isEnum()) {
+				if (!this.requestor.isIgnored(CompletionProposal.FIELD_REF)) {
+					this.assistNodeIsEnum = true;
+					findEnumConstantsFromSwithStatement(this.completionToken, switchStatement);
+				}
+			} else if (exprType.isClass() || exprType.isInterface() || exprType.isRecord()) {
+				// suggest type reference completions for now if its java 21 and above or it is a string
+
+				// We can later suggest more smart completions such as
+				// - completion for Subtypes
+				// - completion for subtypes with variable name, <<TypeName varName>>
+				// - completions for extracing record components a.k.a Record desctructions
```

Comment:
If this is to be read as a TODO list, you may want to update it vis-a-vis your additions here